# Heroes Of Pymoli
### Introductory Data Analysis project using Python Programming

Welcome to the Heroes of Pymoli! This is a fantasy world (much like World of Warcraft) where users create avatars, fight monsters, level up, and, most importantly, gear up to fight around the world of Pymoli. However... Some players are a bit more serious about the game than others. 

Below is my data analysis of certains trends associated with the game. I answer questions like "Which gender is more likely to spend money on the game?" "How much are players willing to spend on a specific item?" & "What age is the majority of the games audience?"

Ideally, the skills shown in this project would be used for data wrangling, mining, and visualization for companies that need a direction in marketing campaigns.

In [113]:
import pandas as pd
import os, csv
import numpy as np

In [114]:
csvpath = os.path.join('Resources/purchases.csv')
dataframe = pd.read_csv(csvpath)
dataframe.head(10)

Purchase ID             SN  Age                 Gender  Item ID  \
0            0        Lisim78   20                   Male      108   
1            1    Lisovynya38   40                   Male      143   
2            2     Ithergue48   24                   Male       92   
3            3  Chamassasya86   24                   Male      100   
4            4      Iskosia90   23                   Male      131   
5            5        Yalae81   22                   Male       81   
6            6      Itheria73   36                   Male      169   
7            7    Iskjaskst81   20                   Male      162   
8            8      Undjask33   22                   Male       21   
9            9    Chanosian48   35  Other / Non-Disclosed      136   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44  
5                                  Dreamkiss   3.61  
6     Interrogator, Blood Blade of the Queen   2.18  
7                              Abyssal Shard   2.67  
8                                  Souleater   1.10  
9               Ghastly Adamantite Protector   3.58

## I. What is the total number of purchases in the dataset?

In [115]:
total_purchases = len(dataframe["Purchase ID"].value_counts())
total_purchases_df = pd.DataFrame([total_purchases], columns = ["TOTAL PURCHASES"])
total_purchases_df

TOTAL PURCHASES
0              780

This data set spans 780 purchases. The total amount of purchases should be displayed in the analysis as a base.

## II. How many players are there in our current data set?

In [116]:
players = len(dataframe["SN"].value_counts())
totalplayers = pd.DataFrame([players], columns = ["TOTAL PLAYERS"])
totalplayers.head()

TOTAL PLAYERS
0            576

Although the CSV we uploaded has 770+ transactions, the total amount of players that purchased item's is only 576. This is because some players were repeat customers. Because Pymoli experiences repeat customers, we need to count the unique values of "SN", the player's unique username.

## III. Out of all the items within the game, how many did players actually purchase?

In [117]:
purchaseditems = len(dataframe["Item Name"].value_counts())
purchaseditems_df = pd.DataFrame([purchaseditems], columns = ["Unique Items Purchased"])
purchaseditems_df

Unique Items Purchased
0                     179

There are a total of 179 unique items that were purchased by players. Here we use the same concept in question I. There are numerous items in the game, but other items are purchased more frequently than others. Therefore, we would need to count the number of unique items in the entire dataset. 

## IV. What is the average amount spent on a single item?

In other words, what is the mean amount that the creators of Heroes of Pymoli make off of a single purchase?

In [118]:
average = dataframe["Price"].mean()
averageprice = round(average, 2)
averageprice_df = pd.DataFrame([averageprice], columns = ["Average Price"])
averageprice_df

Average Price
0           3.05

## V. What was the Total Revenue indicated in this dataset?

In [119]:
revenue = dataframe["Price"].sum()
totalrevenue = pd.DataFrame([revenue], columns = ["Total Revenue"])
totalrevenue

Total Revenue
0        2379.77

Total Revenue is calculated by adding up all the prices spent for each purchase. A simple line of code that shows the total amount gained based off of purchases. Revenue is $2,379.77

## Display the Overall Analysis of all questions answered from I - V

In [120]:
analysis = pd.DataFrame({"TOTAL PLAYERS": [players],
                        "UNIQUE ITEMS PURCHASED": [purchaseditems],
                        "AVERAGE PRICE": [averageprice],
                        "TOTAL PURCHASES": [total_purchases],
                        "TOTAL REVENUE": [revenue]})
analysis.head()

TOTAL PLAYERS  UNIQUE ITEMS PURCHASED  AVERAGE PRICE  TOTAL PURCHASES  \
0            576                     179           3.05              780   

   TOTAL REVENUE  
0        2379.77

## Lets display the master dataframe again to have a reference from this point onward

In [121]:
dataframe.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## VI. What are the gender demographics of this dataset?

In order to thin the marketing audience, we may want to determine things like gender, age, location, etc. Here we start by looking at which gender tends to spend money on the game.

In [122]:
gender = dataframe["Gender"].value_counts()
gender

Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [123]:
male_perc = 652 / 780
female_perc = 113 / 780
non_perc = 15 / 780

male_perc = (format(male_perc, ".2%"))
female_perc = (format(female_perc, ".2%"))
non_perc = (format(non_perc, ".2%"))

print("Male: " + male_perc)
print("Female: " + female_perc)
print("Non-Disclosed: " + non_perc)

Male: 83.59%
Female: 14.49%
Non-Disclosed: 1.92%


In [124]:
genderdemo = pd.DataFrame({
    "Male": [male_perc],
    "Female": [female_perc],
    "Non-Disclosed": [non_perc]
})
genderdemo

Male  Female Non-Disclosed
0  83.59%  14.49%         1.92%

About 84% of the total purchases were made by males. Females accounted for about 14.5% of the total purchases, and almost 2% of the purchases decided not to disclose their gender. 

## Fixing the Gender Demographics

As mentioned earlier, some of the purchases could have been repeat customers. In this case the gender demographics dataframe can be unreliable. We can make the demographics analysis more accurate by removing duplicates before calling in the dataframe.

##### When reformatting a dataframe, it is best to separate it from the master dataframe used in the begging of this project

In [125]:
gender = dataframe[["SN", "Gender"]]
gender.head(10)

SN                 Gender
0        Lisim78                   Male
1    Lisovynya38                   Male
2     Ithergue48                   Male
3  Chamassasya86                   Male
4      Iskosia90                   Male
5        Yalae81                   Male
6      Itheria73                   Male
7    Iskjaskst81                   Male
8      Undjask33                   Male
9    Chanosian48  Other / Non-Disclosed

##### Now we can start dropping the duplicates without changing anything in the master dataframe

In [126]:
gender = gender.drop_duplicates()

In [127]:
gender_df = gender["Gender"].value_counts()
gender_df

Male                     484
Female                    81
Other / Non-Disclosed     11
Name: Gender, dtype: int64

##### Let's create the new demographics dataframe using the updated information above

In [128]:
male_perc = 484 / 576
female_perc = 81 / 576
non_perc = 11 / 576

male_perc = (format(male_perc, ".2%"))
female_perc = (format(female_perc, ".2%"))
non_perc = (format(non_perc, ".2%"))

print("Male: " + male_perc)
print("Female: " + female_perc)
print("Non-Disclosed: " + non_perc)

Male: 84.03%
Female: 14.06%
Non-Disclosed: 1.91%


In [129]:
genderdemo = pd.DataFrame({
    "Male": [male_perc],
    "Female": [female_perc],
    "Non-Disclosed": [non_perc]
})
genderdemo

Male  Female Non-Disclosed
0  84.03%  14.06%         1.91%

In [135]:
totalgenderdemo = [gender_df[0], gender_df[1], gender_df[2]]
totalgenderperc = [male_perc, female_perc, non_perc]

genderdemo_df = pd.DataFrame({
    "TOTAL COUNT PER GENDER": totalgenderdemo,
    "TOTAL PERCENTAGE PER GENDER": totalgenderperc
})

genderdemo_df.index = (["Male", "Female", "Non-Disclosed"])

genderdemo_df

TOTAL COUNT PER GENDER TOTAL PERCENTAGE PER GENDER
Male                              484                      84.03%
Female                             81                      14.06%
Non-Disclosed                      11                       1.91%

I made the Demographics dataframe slightly more accurate by removing the duplicate purchases. For total purchases, companies may want to keep the duplicates in the dataframe to reflect the total number. However, when choosing a target audience (i.e. marketing campaigns), companies may want to remove duplicates.

## Lets display the master dataframe again to have a reference from this point onward

In [136]:
dataframe.head(10)

Purchase ID             SN  Age                 Gender  Item ID  \
0            0        Lisim78   20                   Male      108   
1            1    Lisovynya38   40                   Male      143   
2            2     Ithergue48   24                   Male       92   
3            3  Chamassasya86   24                   Male      100   
4            4      Iskosia90   23                   Male      131   
5            5        Yalae81   22                   Male       81   
6            6      Itheria73   36                   Male      169   
7            7    Iskjaskst81   20                   Male      162   
8            8      Undjask33   22                   Male       21   
9            9    Chanosian48   35  Other / Non-Disclosed      136   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44  
5                                  Dreamkiss   3.61  
6     Interrogator, Blood Blade of the Queen   2.18  
7                              Abyssal Shard   2.67  
8                                  Souleater   1.10  
9               Ghastly Adamantite Protector   3.58

## VII. What are the specifics regarding gender demographics and the total revenue?

To get a more detailed report on the relationship between gender the purchases, I analyzed the average each gender spent, the total that each gender spent, and the average each gender spent in their own bracket.

In [137]:
group_gender = dataframe.groupby(["Gender"])

The above line of code will restructure the output of data to take into account the category "Gender"

In [139]:
average_per_gender = group_gender["Price"].mean()
average_per_gender = round(average_per_gender, 2)
average_per_gender

Gender
Female                   3.20
Male                     3.02
Other / Non-Disclosed    3.35
Name: Price, dtype: float64

The average that males / female / non-disclosed spent on a single item is 3.02 / 3.20 / 3.35 respectively

In [141]:
total_per_gender = group_gender["Price"].sum()
total_per_gender

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

The total that males / females / non-disclosed spent on a single item is 1,967.64 / 361.94 / 50.19 respectively.

In [144]:
average_in_bracket = total_per_gender / gender_df
average_in_bracket = round(average_in_bracket, 2)
average_in_bracket

Female                   4.47
Male                     4.07
Other / Non-Disclosed    4.56
dtype: float64

The average that males spent on a single purchase is 4.07. Females spent on average 4.47 per purchase, and non-disclosed players spent 4.56 on a single item.